content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+```
+
+The `-g` flag tells nvcc to include debugging information for the host code
+(the code that will run on the CPU) and the `-G` flag tells nvcc to include
+debugging information for the device code (the code that will run on the GPU).
+`-O0` disables optimizations for the host code. `-G` implicitly disables
+optimizations for the device code.
+
+`-c` tells the compiler to produce an object file instead of a executable. An
+object file is a mostly-compiled program that only needs to be linked to create
+an executable. After you compile the .cpp file in the next section, you will
+link the object files together to create the final executable.
+
+nvcc does not actually fully compile C/C++. It actually processes the CUDA
+directives in the .cu file and produces a new C/C++ file where the CUDA
+directives are replaced with the proper C/C++ code that interacts with the CUDA
+API to do the work of loading the correct GPU driver etc. `-ccbin=cc` tells
+nvcc to use the `cc` compiler to do the actual compilation of the generated
+C/C++ code. When the Cray CCE module is loaded, cc is the Cray C/C++ compiler.
+
+`-gencode` tells nvcc which GPU architecture to target. The actual value for
+this argument will vary based on which GPU you are targeting. Refer to the
+[NVIDIA compute capabilities table](https://developer.nvidia.com/cuda-gpus) to
+find the compute capability of your GPU. For example, the A100 GPU that I am
+targeting while writing this tutorial has a compute capability of 8.0. Take the compute
+capability of your GPU and pass it to nvcc via the `-gencode` flag by
+removing the decimal point and embedding it in an `arch=compute_XX,code=sm_XX`
+pair. For my A100 example, the complete option is `-gencode compute_80,code=sm_80`.
+
+`simpleMI.cu` is the input file and `-o simpleMPI_gpu.o` tells `nvcc` to call
+the output file `simpleMPI_gpu.o`.
+
+My complete nvcc invocation for an A100 GPU looks like this:
+
+```shell
+$ nvcc -g -G -c -ccbin=cc -gencode arch=compute_80,code=sm_80 simpleMPI.cu -o simpleMPI_gpu.o
+```
+
+Refer to the [`nvcc` user manual](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html)
+and the [compilation section of the cuda-gdb documentation](https://docs.nvidia.com/cuda/cuda-gdb/index.html#compiling-the-application)
+for more information.
+
+#### Compile the .cpp File
+
+The simpleMPI.cpp file contains code that will be run on the CPU. It is the
+entry point to the program and contains the MPI operations. Use CC to compile
+it.
+
+```shell
+$ CC -g -O0 -c simpleMPI.cpp -o simpleMPI_cpu.o
+```
+
+The flags have the same meanings as in the nvcc compilation. The Cray CC
+compiler will automatically include the flags needed for building with MPI.
+
+#### Link the Application and Produce an Executable
+
+Finally, use CC to link the two object files into the final executable.
+
+```shell
+$ CC simpleMPI_gpu.o simpleMPI_cpu.o -o simpleMPI
+```
+
+#### Test The Application
```

Comment:
"Test the application" (even if you retained title case, "the" would always have a lower case "t")